src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -475,14 +490,98 @@ InsertShardPlacementRows(Oid relationId, int64 shardId, List *workerNodeList,
 
 /*
  * CreateShardsOnWorkers creates shards on worker nodes given the shard placements
- * as a parameter. Function opens connections in transactional way. If the caller
- * needs an exclusive connection (in case of distributing local table with data
- * on it) or creating shards in a transaction, per placement connection is opened
- * for each placement.
+ * as a parameter. Function branches into two: either use the executor or execute the
+ * commands one by one.
  */
 void
 CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
 					  bool useExclusiveConnection, bool colocatedShard)
+{
+	if (DEFAULT_POOL_SIZE > 0)
+	{
+		CreateShardsOnWorkersViaExecutor(distributedRelationId, shardPlacements,
+										 useExclusiveConnection, colocatedShard);
+	}
+	else
+	{
+		CreateShardsOnWorkersViaCommands(distributedRelationId, shardPlacements,
+										 useExclusiveConnection, colocatedShard);
+	}
+}
+
+
+/*
+ * CreateShardsOnWorkersViaExecutor creats the shards via the executor.
+ */
+static void
+CreateShardsOnWorkersViaExecutor(Oid distributedRelationId, List *shardPlacements, bool
+								 useExclusiveConnection, bool colocatedShard)
+{
+	bool includeSequenceDefaults = false;
+	List *ddlCommandList = GetTableDDLEvents(distributedRelationId,
+											 includeSequenceDefaults);
+	List *foreignConstraintCommandList =
+		GetTableForeignConstraintCommands(distributedRelationId);
+	ListCell *shardPlacementCell = NULL;
+
+	int taskId = 1;
+	List *taskList = NIL;
+
+	int poolSize = useExclusiveConnection ? DEFAULT_POOL_SIZE : 1;
+
+	foreach(shardPlacementCell, shardPlacements)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(shardPlacementCell);
+		uint64 shardId = shardPlacement->shardId;
+		ShardInterval *shardInterval = LoadShardInterval(shardId);
+		int shardIndex = -1;
+		List *commandList = NIL;
+		Task *task = NULL;
+		List *relationShardList = RelationShardListForShardCreate(shardInterval);
+
+		if (colocatedShard)
+		{
+			shardIndex = ShardIndex(shardInterval);
+		}
+
+		/* we only need to calculate this once per shardInterval, not placement */
+		if (relationShardList == NIL)
```

Comment:
This can be removed. `relationShardList` has already been created earlier.

Reply:
removed in the other PR. I'll not rebase until we merge the other PR because it'd be a complicated rebase.